### ISSUE TITLE
Constant folding or simplification in the hybrid evaluator

### ISSUE THREAD

**Issue by krlmlr** [hadley/dplyr#2143]
Just tried:

``` r
mtcars %>% mutate(few_cyls = cyl %in% c(4, 6))
```

Even with hybrid handling of `%in%` (#126) this is currently offloaded to R, because `c(4, 6)` is not a single variable. Other handlers (min/max, ...) could also benefit.

`compiler:::constantFold()` isn't exported yet, unfortunately.

**Comment by krlmlr** [comment]
But we can leverage `compiler::compile()`:

```r
compiler::compile(quote(T), options = list(optimize = 3)) %>% eval(new.env(parent = emptyenv()))
```

Just need to figure out which functions in the environment are pure, so that the `c()` example from above works.

**Comment by hadley** [comment]
Seems like a reasonable approach to try

**Comment by krlmlr** [comment]
Current road map:

- [ ] Merge #2309
- [ ] Split `GroupedHybridCall` class (krlmlr#20)
- [ ] Walk the parse tree only once in `GroupedHybridCall` (https://github.com/hadley/dplyr/pull/2298#issuecomment-266047672), replace subtrees by XPtr-s that hold the `Result` object which performs hybrid evaluation
- [ ] When evaluating groups, call the evaluators and replace the XPtr-s again by the real values
- [ ] Add constant folding and better argument matching to the hybrid evaluators (#2298)

**Comment by lionel-** [comment]
I wonder if we should just drop support for partial matching in dplyr to simplify argument matching? This way matching becomes a simple linear search through arglist and dots.

Would it make sense to rename the `Result` class to `Hybrid` or something more suggestive?

One issue with replacing subtrees by XPtr is that you cannot nest hybrid calls.

**Comment by krlmlr** [comment]
We currently don't support partial matching or argument reordering. I thought we could use match.call() to solve both, but there are certainly other options. Constant folding is the other expensive part, we want to do this only once (and not once per group). Argument matching and constant folding are somewhat orthogonal, but both require a preprocessing stage that performs c.f./a.m. separately from the actual evaluation. And then we need a way to store the preprocessed result: instead of storing an XPtr, we could patch the call to insert constants and reorder arguments. I'll update the road map.

Renaming: Certainly possible, but is a big change with needs to be done when all PRs and branches are closed.

**Comment by krlmlr** [comment]
The XPtr has one advantage, though: Argument matching and creation of the `Result` object can be performed entirely at the preprocessing stage. I'd argue that this is faster than performing the same argument matching all over again for each group. Need to measure the actual costs of the argument matching in a "typical" scenario.

**Comment by lionel-** [comment]
How about this strategy: traverse the call tree only once, and save a sequentialisation of the path when an hybridable call is found. They could be saved in a `std::vector` of `std::pair` of tree traces and `Result` objects. You keep on doing this until the end of the tree. Then during group evaluation the results would be inlined in the call (the group's duplicate), in reverse order.

This way `Result` objects are initialised only once and nested hybrid calls are properly supported. Also this avoids traversing calls for each group to look for xptrs in case no hybrid call was found.

**Comment by krlmlr** [comment]
Yes, that's even better. Either way, I think it's very important that the logic is properly encapsulated, I don't want to be handling sequences of `CDR` and `CAR` calls in the hybrid evaluator. We should have a `CallTreeMap` class that maps nodes in the call tree to `Result` pointers.

The hybrid evaluator needs to know the data type at parse time. So, to support nesting, we may need to look in the `CallTreeMap`.

**Comment by krlmlr** [comment]
@lionel-: I wonder if the `CallTreeMap` (or whatever it's called) should live in `rlang`.

**Comment by lionel-** [comment]
What would this class do exactly? The thing is that rlang should have no dependencies at all, so it can only have C code.

**Comment by krlmlr** [comment]
The CallTreeMap would be able to walk the tree, record a path to the current node, and later substitute entries at recorded paths with user-provided SEXPs. Would this be useful outside of dplyr?

`rlang` might be able to use C++, but not Rcpp according to your scenario. (One option is to embed a copy of Rcpp in rlang, but then we'd need to GPL rlang.) Anyway, we can draft the code in a separate package first (so that it can be tested easier) and later decide where it should go.

**Comment by lionel-** [comment]
I would rather move away from Rcpp which I dislike because it kills compilation time, has a lot of unnecessary stuff, and focuses on the wrong idioms (tries to be too R-like) :)

It'd be nice if we could recast the CallTreeMap problem into an iterator-based idiom. But I think walking trees with iterators is a difficult problem.

Similarly, I've been thinking about replacing the SlicingIndex stuff with subsequences:

- NaturalSlicingIndex: list of 1 subsequence 1:N (actually the original sequence)
- OffsetSlicingIndex:  list of 1 subsequence a:b (incongruous because the only one that isn't a partition)
- GroupedSlicingIndex: list of J subsequences of length `n_j`, likely uncontiguous.
- RowwiseSlicingIndex: list of N subsequences of length 1

You could request subsequences as a partition of the R vector, or you could request a particular subsequence from a particular group.

Each subsequence would be a simple wrapper with `begin()` and `end()` methods. Then you'd map the subsequences like this for example:

```c++
std::vector subseqs { seq1, seq2 };

for (auto& seq : subseqs) {
  std::accumulate(seq.begin(), seq.end(), hybrid_f{})
}
```

The iterators of subsequences would handle non-contiguity themselves.

**Comment by lionel-** [comment]
> rlang might be able to use C++, but not Rcpp according to your scenario.

yes but we're currently stuck because we need to extract at the minimum all the exception-catching and maybe also the attributes/exporting stuff from Rcpp to be able to effectively use C++ :/

**Comment by krlmlr** [comment]
Yeah, let's design a nice interface to `CallTreeMap`, iterator-based or not. Mind that we can't use `auto` yet because we're stuck at C++98.

Let's discuss the SlicingIndex elsewhere, perhaps in vctrs? My feeling is that the SlicingIndex is just fine (perhaps doesn't need to know its group ID) but we need a container of SlicingIndices which can optionally be a partition (think joins).

**Comment by lionel-** [comment]
> we're stuck at C++98

Couldn't we move to 11? At least for vctrs I'd like to use c++11.

> Let's discuss the SlicingIndex elsewhere, perhaps in vctrs?

Yes I've been thinking about your PR a lot, I'll come back to you soon. We should probably meet in a video call after the holidays to discuss all this.

**Comment by lionel-** [comment]
will you be there at the RStudio conf?

**Comment by krlmlr** [comment]
According to @hadley, we can't move to C++11 with dplyr until Red Hat provides a compatible compiler by default, see also #2230. According to their [timeline](https://access.redhat.com/support/policy/updates/errata#Life_Cycle_Dates), it's 2017 or 2020 or even later for RHEL 5/6, depending on which production phase we support. Only [RHEL 7 has GCC 4.8](https://access.redhat.com/solutions/19458).

If we use C++11 in vctrs, dplyr and other packages with similar restrictions can't use it until then.

I can't make it to the RStudio conf this year, but I'm happy to discuss over VC.

**Comment by krlmlr** [comment]
Closing in favor of #2299.
